107-tee-boot-attestation.md

```diff
@@ -0,0 +1,360 @@
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+-->
+
+Trusted Execution Environments (TEEs) are a confidential computing and
+virtualization technology that allow for the protection/encryption of guest VM
+RAM, cache memory, and CPU registers for running sensitive applications on
+potentially untrusted hosts within the cloud/edge. With TEEs, data written to
+RAM, caches, or registers are encrypted with keys maintained by secure
+processors embedded within a CPU, thus protecting confidential VM guests from
+other guests on a host, or even the host system itself. With encryption fully
+managed by the hardware, only the confidential VM (CVM) itself would be able to
+read/write to its own memory, protecting against buggy/malicious hosts spying
+on or tampering with the CVM.
+
+TEEs are for protecting applications/users on untrusted platforms. In-fact, the
+confidential computing threat model assumes that the host in which a CVM is
+running on top of is actively attempting to tamper or spy on guest memory. With
+this, users must *ensure* that a host has launched their guest with all proper
+TEE protections and didn't perform any other nefarious actions to tamper with
+a CVM. That is, CVMs must *attest* their boot environment and prove the
+*integrity* of their workload before any sensitive operations can be performed
+within a CVM.
+
+This enhancement will introduce a new handler to the keylime verifier strictly
+for performing TEE boot attestation. However, with untrusted hosts controlling
+guest memory, certain attacks can take place *before* a guest OS is booted and a
+keylime agent has been initialized. Timing of boot attestation is critical,
+and thus a TEE boot attestation handler cannot assume that it is communicating
+with a keylime agent. Furthermore, the TEE handler will release some secret
+(whether that is a public key to extend an initial TPM measurement with, a LUKS
+key to unlock some initial TPM state, or a combination of both) upon a
+successful boot attestation that will be required for guest OS to boot or a
+keylime agent to start running. In-effect, this handler would essentially
+bootstrap a keylime agent running in a CVM, as the TEE boot environment would
+need to be validated before a keylime agent could begin running.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this enhancement.  Describe why the change is important and the benefits to users.
+-->
+
+As CVMs are deployed on untrusted systems, it is reasonable to assume that a
+CVM user would also like to take advantage of system integrity monitoring
+provided by keylime *in addition to* TEE technology. That is, it is reasonable
+to expect that keylime agents will be deployed on CVMs. With that, it is
+reasonable to include TEE attestation mechanisms for environments in which
+keylime will already be running for system integrity monitoring purposes.
```

Comment:
I certainly agree that `Keylime` should support CVMs, and actually an initial crude prototype was done @ruocco (never released, but tested internally at IBM), where the `registrar` was modified to process the "attestation report" for the (then, just SEV) guest, in order to decide if it would accept the EK from that particular `agent`. 

But it is important to avoid conflating the two issues: a) `Keylime` should incorporate mechanisms to take into account "boot artifacts" from "TEE technology" in order to decide if a node (i.e., CVM) is "attested" and b) how to implement this clearly necessary/welcome/desired functionality. At this point allow me to reiterate the general `Keylime` philosophy: the attestation is performed after the fact, and it does not step in to prevent a node from getting into an "non-attested" state.